File: apps/android/app/src/main/java/com/nanolink/app/ui/screens/DashboardScreen.kt
package com.nanolink.app.ui.screens

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.List
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.Computer
import androidx.compose.material.icons.outlined.Dashboard
import androidx.compose.material.icons.outlined.Memory
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material.icons.outlined.Storage
import androidx.compose.material.icons.outlined.Warning
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nanolink.app.data.model.Agent
import com.nanolink.app.data.model.AuditEntry
import com.nanolink.app.state.AppViewModel
import com.nanolink.app.ui.design.NanoCard
import com.nanolink.app.ui.design.NanoEmptyState
import com.nanolink.app.ui.design.NanoKpiTile
import com.nanolink.app.ui.design.NanoListRow
import com.nanolink.app.ui.design.NanoMonoFamily
import com.nanolink.app.ui.design.NanoSectionLabel
import com.nanolink.app.ui.design.NanoSparkline
import com.nanolink.app.ui.design.NanoStatusDot
import com.nanolink.app.ui.design.nano
import com.nanolink.app.ui.tr
import kotlin.math.roundToInt

@Composable
fun DashboardScreen(viewModel: AppViewModel, modifier: Modifier = Modifier) {
    val t = nano
    val servers by viewModel.servers.collectAsStateWithLifecycle()
    val agents by viewModel.agents.collectAsStateWithLifecycle()
    val metrics by viewModel.metrics.collectAsStateWithLifecycle()
    val activity by viewModel.activity.collectAsStateWithLifecycle()
    val activeServerId by viewModel.activeServerId.collectAsStateWithLifecycle()

    val online = agents.filter { it.isOnline }
    val offline = agents.filter { !it.isOnline }

    fun average(select: (com.nanolink.app.data.model.AgentMetrics) -> Double): Double {
        val values = online.mapNotNull { metrics[it.id] }.map(select)
        return if (values.isEmpty()) 0.0 else values.average()
    }

    val avgCpu = average { it.cpuPercent }
    val avgMemory = average { it.memoryPercent }
    val memoryUsedGiB = online.mapNotNull { metrics[it.id]?.memory?.used }.sum() / (1024.0 * 1024.0 * 1024.0)
    val diskAlerts = agents.count { agent ->
        metrics[agent.id]?.disks?.any { it.usagePercent > 85 } == true
    }
    val topCpu = online.sortedByDescending { metrics[it.id]?.cpuPercent ?: 0.0 }.take(4)
    val recentActivity = activity.values.flatten().sortedByDescending { it.atMillis }.take(5)

    var cpuSpark by remember { mutableStateOf<List<Double>>(emptyList()) }
    var memSpark by remember { mutableStateOf<List<Double>>(emptyList()) }

    LaunchedEffect(metrics, activeServerId) {
        val samples = 20
        cpuSpark = List(samples) { avgCpu }
        memSpark = List(samples) { avgMemory }
    }

    if (servers.isEmpty()) {
        Box(modifier = modifier.fillMaxSize().padding(16.dp), contentAlignment = Alignment.Center) {
            NanoEmptyState(
                icon = Icons.Outlined.Dashboard,
                title = tr("dashboard.noServersYet"),
                detail = tr("dashboard.addFirstServer"),
            )
        }
        return
    }

    Box(modifier = modifier.fillMaxSize()) {
        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            contentPadding = PaddingValues(start = 16.dp, end = 16.dp, top = 8.dp, bottom = 110.dp),
            verticalArrangement = Arrangement.spacedBy(0.dp),
        ) {
            item {
                DashboardHeader(servers.size, activeServerId)
                Spacer(Modifier.height(12.dp))
            }

            if (servers.size > 1) {
                item {
                    ServerChips(servers.map { it.name }, activeServerId ?: "")
                    Spacer(Modifier.height(12.dp))
                }
            }

            item {
                KpiGrid(
                    avgCpu = avgCpu,
                    avgMemory = avgMemory,
                    memoryUsedGiB = memoryUsedGiB,
                    onlineCount = online.size,
                    totalCount = agents.size,
                    diskAlerts = diskAlerts,
                    cpuSpark = cpuSpark,
                    memSpark = memSpark,
                )
                Spacer(Modifier.height(12.dp))
            }

            if (offline.isNotEmpty()) {
                item {
                    OfflineBanner(offline.size)
                    Spacer(Modifier.height(12.dp))
                }
            }

            item {
                NanoSectionLabel(tr("dashboard.topCpu"))
            }

            if (topCpu.isEmpty()) {
                item {
                    NanoCard(modifier = Modifier.padding(vertical = 12.dp)) {
                        Box(
                            modifier = Modifier.fillMaxWidth().padding(vertical = 22.dp),
                            contentAlignment = Alignment.Center,
                        ) {
                            Text(
                                tr("dashboard.noOnlineNodes"),
                                fontSize = 13.sp,
                                color = t.fg4,
                            )
                        }
                    }
                }
            } else {
                item {
                    NanoCard {
                        topCpu.forEachIndexed { index, agent ->
                            TopCpuRow(agent, metrics[agent.id], divider = index < topCpu.size - 1)
                        }
                    }
                }
            }

            item {
                Spacer(Modifier.height(12.dp))
                NanoSectionLabel(tr("dashboard.recentActivity"))
            }

            if (recentActivity.isEmpty()) {
                item {
                    NanoCard(modifier = Modifier.padding(vertical = 12.dp)) {
                        Box(
                            modifier = Modifier.fillMaxWidth().padding(vertical = 22.dp),
                            contentAlignment = Alignment.Center,
                        ) {
                            Text(tr("dashboard.noActivity"), fontSize = 13.sp, color = t.fg4)
                        }
                    }
                }
            } else {
                item {
                    NanoCard {
                        recentActivity.forEachIndexed { index, entry ->
                            AuditRow(entry, divider = index < recentActivity.size - 1)
                        }
                    }
                }
            }
        }

        // Floating action button with gradient and shadow.
        FloatingActionButton(
            onClick = { /* Navigate to add server */ },
            containerColor = t.accent,
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(end = 16.dp, bottom = 16.dp),
        ) {
            Row(
                modifier = Modifier.padding(horizontal = 18.dp, vertical = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Icon(Icons.Outlined.Add, contentDescription = null, tint = t.onAccent)
                Text(
                    tr("dashboard.newNode"),
                    fontSize = 15.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = t.onAccent,
                )
            }
        }
    }
}

@Composable
private fun DashboardHeader(serverCount: Int, activeServerId: String?) {
    val t = nano
    Row(
        modifier = Modifier.fillMaxWidth().padding(top = if (t.isIOS) 32.dp else 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(
            text = tr("dashboard.title"),
            fontSize = if (t.isIOS) 32.sp else 28.sp,
            fontWeight = if (t.isIOS) FontWeight.Bold else FontWeight.SemiBold,
            letterSpacing = if (t.isIOS) (-0.6).sp else 0.sp,
            color = t.fg,
        )
        Spacer(Modifier.weight(1f))
        CircleButton(icon = Icons.Outlined.Settings, gradient = true) { /* Assistant */ }
        CircleButton(icon = Icons.AutoMirrored.Outlined.List) { /* Agents */ }
        CircleButton(icon = Icons.Outlined.Computer) { /* Server switch */ }
    }
}

@Composable
private fun CircleButton(icon: ImageVector, gradient: Boolean = false, onClick: () -> Unit) {
    val t = nano
    Box(
        modifier = Modifier
            .size(34.dp)
            .clip(CircleShape)
            .background(
                if (gradient) {
                    Brush.linearGradient(listOf(t.accent, t.tertiary))
                } else {
                    Brush.linearGradient(listOf(t.card, t.card))
                },
            )
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            icon,
            contentDescription = null,
            tint = if (gradient) t.onAccent else t.accent,
            modifier = Modifier.size(16.dp),
        )
    }
}

@Composable
private fun ServerChips(names: List<String>, activeId: String) {
    val t = nano
    LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        items(names) { name ->
            val active = name.hashCode().toString() == activeId
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(16.dp))
                    .background(if (active) t.accent.copy(alpha = 0.15f) else t.card2)
                    .clickable { /* Switch server */ }
                    .padding(horizontal = 14.dp, vertical = 8.dp),
            ) {
                Text(
                    name,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Medium,
                    color = if (active) t.accent else t.fg2,
                )
            }
        }
    }
}

@Composable
private fun KpiGrid(
    avgCpu: Double,
    avgMemory: Double,
    memoryUsedGiB: Double,
    onlineCount: Int,
    totalCount: Int,
    diskAlerts: Int,
    cpuSpark: List<Double>,
    memSpark: List<Double>,
) {
    val t = nano
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            NanoKpiTile(
                label = tr("dashboard.kpi.cpu"),
                value = "${avgCpu.roundToInt()}%",
                sub = tr("dashboard.kpi.avgAll"),
                icon = Icons.Outlined.Memory,
                tone = if (avgCpu > 85) "crit" else if (avgCpu > 70) "warn" else null,
                spark = cpuSpark,
                modifier = Modifier.weight(1f),
            )
            NanoKpiTile(
                label = tr("dashboard.kpi.memory"),
                value = "${avgMemory.roundToInt()}%",
                sub = String.format("%.1f GiB " + tr("dashboard.kpi.used"), memoryUsedGiB),
                icon = Icons.Outlined.Dashboard,
                tone = if (avgMemory > 85) "crit" else if (avgMemory > 75) "warn" else null,
                spark = memSpark,
                modifier = Modifier.weight(1f),
            )
        }
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            NanoKpiTile(
                label = tr("dashboard.kpi.nodes"),
                value = "$onlineCount/$totalCount",
                sub = tr("dashboard.kpi.online"),
                icon = Icons.Outlined.Computer,
                modifier = Modifier.weight(1f),
            )
            NanoKpiTile(
                label = tr("dashboard.kpi.diskAlerts"),
                value = diskAlerts.toString(),
                sub = tr("dashboard.kpi.over85"),
                icon = Icons.Outlined.Storage,
                tone = if (diskAlerts > 0) "warn" else null,
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun OfflineBanner(count: Int) {
    val t = nano
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(t.cardRadius))
            .background(t.warn.copy(alpha = 0.12f))
            .padding(horizontal = 14.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(Icons.Outlined.Warning, contentDescription = null, tint = t.warn, modifier = Modifier.size(18.dp))
        Text(
            tr("dashboard.offlineNodes", "count" to count),
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
            color = t.warn,
        )
    }
}

@Composable
private fun TopCpuRow(
    agent: Agent,
    metrics: com.nanolink.app.data.model.AgentMetrics?,
    divider: Boolean,
) {
    val t = nano
    val cpu = metrics?.cpuPercent ?: 0.0
    NanoListRow(divider = divider) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            NanoStatusDot(color = if (agent.isOnline) t.ok else t.fg4, pulse = agent.isOnline)
            Spacer(Modifier.width(12.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text(agent.hostname, fontSize = 14.sp, fontWeight = FontWeight.Medium, color = t.fg)
                Text("${agent.os} ${agent.arch}", fontSize = 11.sp, fontFamily = NanoMonoFamily, color = t.fg4)
            }
            Spacer(Modifier.width(12.dp))
            Column(horizontalAlignment = Alignment.End) {
                Text(
                    "${cpu.roundToInt()}%",
                    fontSize = 15.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = if (cpu > 85) t.crit else if (cpu > 70) t.warn else t.fg,
                )
                Text("CPU", fontSize = 10.sp, fontFamily = NanoMonoFamily, color = t.fg4)
            }
        }
    }
}

@Composable
private fun AuditRow(entry: AuditEntry, divider: Boolean) {
    val t = nano
    NanoListRow(divider = divider) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                when (entry.type.uppercase()) {
                    "CREATE", "START" -> Icons.Outlined.Check
                    "DELETE", "STOP" -> Icons.Outlined.Close
                    else -> Icons.Outlined.Settings
                },
                contentDescription = null,
                tint = t.fg3,
                modifier = Modifier.size(14.dp),
            )
            Spacer(Modifier.width(10.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text("${entry.type} ${entry.target}", fontSize = 13.sp, color = t.fg2, maxLines = 1)
                Text(
                    java.text.SimpleDateFormat("HH:mm:ss", java.util.Locale.US)
                        .format(java.util.Date(entry.atMillis)),
                    fontSize = 10.sp,
                    fontFamily = NanoMonoFamily,
                    color = t.fg4,
                )
            }
        }
    }
}

File: apps/ios/NanoOps/Screens/DashboardScreen.swift
import SwiftUI

/// Aggregate overview for the active server: rolling KPIs, offline nodes, top CPU
/// nodes and the most recent audit activity.
struct DashboardScreen: View {
    /// Set by the regular-width shell so node rows select in the Nodes workspace
    /// instead of pushing a detail screen onto this stack.
    var onSelectAgent: ((Agent) -> Void)? = nil

    @EnvironmentObject private var store: AppStore
    @EnvironmentObject private var router: ShellRouter
    @Environment(\.nano) private var t
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass
    @State private var cpuSpark: [Double] = []
    @State private var memSpark: [Double] = []
    @State private var sparkServerId: String?
    @State private var showAddServer = false
    @State private var showServerSwitch = false

    private var agents: [Agent] { store.agentsForServer() }
    private var online: [Agent] { agents.filter(\.isOnline) }
    private var offline: [Agent] { agents.filter { !$0.isOnline } }

    private func average(_ select: (AgentMetrics) -> Double) -> Double {
        let values = online.compactMap { store.metricsFor($0.id) }.map(select)
        return values.isEmpty ? 0 : values.reduce(0, +) / Double(values.count)
    }

    private var avgCpu: Double { average(\.cpuPercent) }
    private var avgMemory: Double { average(\.memoryPercent) }
    private var memoryUsedGiB: Double {
        Fmt.gib(Double(online.compactMap { store.metricsFor($0.id)?.memory.used }.reduce(0, +)))
    }
    private var diskAlerts: Int {
        agents.filter { agent in
            store.metricsFor(agent.id)?.disks.contains { $0.usagePercent > 85 } == true
        }.count
    }
    private var topCpu: [Agent] {
        online.sorted { (store.metricsFor($0.id)?.cpuPercent ?? 0) > (store.metricsFor($1.id)?.cpuPercent ?? 0) }
    }

    var body: some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: 0) {
                // The Mac window titles the section itself and carries the global
                // actions in its toolbar, so the inline header is phone/iPad only.
                if !t.desktop { header }
                serverChips.padding(.top, t.desktop ? 4 : 12)
                kpiGrid.padding(.top, 12)
                if !offline.isEmpty { offlineBanner.padding(.top, 12) }

                NanoSectionLabel(tr("dashboard.topCpu"))
                if topCpu.isEmpty {
                    emptyHint(icon: "server.rack", text: store.servers.isEmpty ? tr("dashboard.noServersYet") : tr("dashboard.noOnlineNodes"))
                } else {
                    NanoCard {
                        ForEach(Array(topCpu.prefix(4).enumerated()), id: \.element.id) { index, agent in
                            let divider = index < min(topCpu.count, 4) - 1
                            if let onSelectAgent = onSelectAgent {
                                Button { onSelectAgent(agent) } label: {
                                    topCpuRow(agent, divider: divider)
                                }.buttonStyle(.plain)
                            } else {
                                NavigationLink(value: AgentRoute(agent: agent)) {
                                    topCpuRow(agent, divider: divider)
                                }.buttonStyle(.plain)
                            }
                        }
                    }
                }

                NanoSectionLabel(tr("dashboard.recentActivity"))
                if store.recentActivity().isEmpty {
                    NanoCard(padding: EdgeInsets(top: 22, leading: 12, bottom: 22, trailing: 12)) {
                        Text(tr("dashboard.noActivity"))
                            .font(.system(size: 13)).foregroundColor(t.fg4)
                            .frame(maxWidth: .infinity)
                    }
                } else {
                    NanoCard {
                        ForEach(Array(store.recentActivity().prefix(5).enumerated()), id: \.element.id) { index, entry in
                            auditRow(entry, divider: index < min(store.recentActivity().count, 5) - 1)
                        }
                    }
                }
            }
            .padding(EdgeInsets(top: 8, leading: 16,
                                bottom: t.desktop ? 24 : (horizontalSizeClass == .regular ? 48 : 110),
                                trailing: 16))
        }
        .nanoPullToRefresh(enabled: !t.desktop) { await refresh() }
        .background(t.bg)
        .navigationDestination(for: AgentRoute.self) { AgentDetailScreen(agent: $0.agent) }
        .navigationDestination(isPresented: $showAddServer) { AddServerScreen() }
        .sheet(isPresented: $showServerSwitch) { ServerSwitchSheet() }
        .toolbar {
            if t.desktop {
                ToolbarItem(placement: .primaryAction) {
                    Button { showAddServer = true } label: {
                        Image(systemName: "plus")
                    }
                    .buttonStyle(NanoToolbarButtonStyle())
                    .accessibilityLabel(tr("dashboard.newNode"))
                    .help(tr("dashboard.newNode"))
                }
            }
        }
        .nanoFloatingAction(enabled: !t.desktop) {
            Button { showAddServer = true } label: {
                Label(tr("dashboard.newNode"), systemImage: "plus")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 18).frame(height: 48)
                    .background(LinearGradient(colors: [t.accent, t.tertiary], startPoint: .leading, endPoint: .trailing))
                    .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .shadow(color: t.accent.opacity(0.3), radius: 8, y: 5)
            }
            .buttonStyle(.plain)
            .padding(.trailing, 16).padding(.bottom, 4)
        }
        .task { await refresh() }
        .onReceive(store.$allMetrics) { _ in pushSparks() }
        .onChange(of: store.activeServerIdRaw) { _ in pushSparks() }
    }

    private var header: some View {
        HStack(spacing: 8) {
            Text(tr("dashboard.title"))
                .font(.system(size: t.isIOS ? 32 : 28, weight: t.displayWeight))
                .tracking(t.displayTracking).foregroundColor(t.fg)
            Spacer()
            NavigationLink { AssistantScreen() } label: {
                circleButton("sparkles", gradient: true)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(tr("nav.assistant"))
            .help(tr("nav.assistant"))
            NavigationLink { AgentsScreen() } label: { circleButton("magnifyingglass") }
                .buttonStyle(.plain)
                .accessibilityLabel(tr("agents.searchHint"))
                .help(tr("agents.searchHint"))
            Button { showServerSwitch = true } label: { circleButton("server.rack") }
                .buttonStyle(.plain)
                .accessibilityLabel(tr("agents.switchServer"))
                .help(tr("agents.switchServer"))
        }
        .padding(.top, t.isIOS ? 32 : 4)
    }

    private func circleButton(_ icon: String, gradient: Bool = false) -> some View {
        ZStack {
            if gradient { LinearGradient(colors: [t.accent, t.tertiary], startPoint: .topLeading, endPoint: .bottomTrailing) }
            else { t.card }
            Image(systemName: icon).font(.system(size: 16)).foregroundColor(gradient ? .white : t.accent)
        }
        .frame(width: 34, height: 34).clipShape(Circle())
    }

    private var serverChips: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(store.servers) { server in
                    Button { store.setActiveServer(server.id) } label: {
                        HStack(spacing: 6) {
                            NanoStatusDot(color: server.isConnected ? t.ok : t.crit, pulse: server.isConnected)
                            Text(server.name).font(.system(size: 13.5, weight: .medium)).foregroundColor(t.fg)
                            Text("\(store.agentsForServer(server.id).count)").font(NanoFont.mono(10.5)).foregroundColor(t.fg4)
                        }
                        .padding(.horizontal, 12).frame(height: 34)
                        .background(server.id == store.activeServerId ? t.card : Color.clear)
                        .overlay(Capsule().stroke(t.sep2, lineWidth: 1))
                        .clipShape(Capsule())
                    }.buttonStyle(.plain)
                }
                Button { showAddServer = true } label: {
                    Label(tr("dashboard.add"), systemImage: "plus")
                        .font(.system(size: 13)).foregroundColor(t.fg3)
                        .padding(.horizontal, 12).frame(height: 34)
                        .overlay(Capsule().stroke(t.sep, lineWidth: 1))
                }.buttonStyle(.plain)
            }
        }
    }

    private var kpiGrid: some View {
        LazyVGrid(columns: Array(repeating: GridItem(.flexible(), spacing: 10),
                                 count: horizontalSizeClass == .regular ? 4 : 2), spacing: 10) {
            NanoKpiTile(label: tr("dashboard.onlineNodes"), sub: tr("dashboard.nodesOffline", ["n": offline.count]), icon: "server.rack") {
                HStack(alignment: .firstTextBaseline, spacing: 0) {
                    Text("\(online.count)")
                    Text("/\(agents.count)").font(.system(size: 18)).foregroundColor(t.fg4)
                }
            }
            NanoKpiTile(label: tr("dashboard.avgCpu"), sub: tr("dashboard.avgCpuLast60s"), icon: "cpu",
                        tone: avgCpu > 80 ? "warn" : nil, spark: cpuSpark.count >= 2 ? cpuSpark : nil) {
                Text(String(format: "%.1f%%", avgCpu))
            }
            NanoKpiTile(label: tr("dashboard.avgMemory"), sub: tr("dashboard.avgMemUsed", ["gib": String(format: "%.0f", memoryUsedGiB)]), icon: "memorychip",
                        tone: avgMemory > 80 ? "warn" : nil, spark: memSpark.count >= 2 ? memSpark : nil) {
                Text(String(format: "%.1f%%", avgMemory))
            }
            NanoKpiTile(label: tr("dashboard.diskAlerts"), sub: tr("dashboard.diskAlertsSub"), icon: "exclamationmark.triangle",
                        tone: diskAlerts > 0 ? "warn" : nil) { Text("\(diskAlerts)") }
        }
    }

    private var offlineBanner: some View {
        Button { router.showNodes(filter: "offline") } label: {
            HStack(alignment: .top, spacing: 10) {
                Image(systemName: "exclamationmark.triangle").foregroundColor(t.crit)
                VStack(alignment: .leading, spacing: 2) {
                    Text(tr("dashboard.offlineBanner", ["n": offline.count])).font(.system(size: 13.5, weight: .medium)).foregroundColor(t.fg)
                    Text(offline.map(\.hostname).joined(separator: " · ")).font(.system(size: 11.5)).foregroundColor(t.fg3).lineLimit(1)
                }
                Spacer()
                Text(tr("dashboard.view")).font(.system(size: 13, weight: .semibold)).foregroundColor(t.crit)
            }
            .padding(12).background(t.crit.opacity(0.1))
            .overlay(RoundedRectangle(cornerRadius: 12).stroke(t.crit.opacity(0.25), lineWidth: 0.5))
            .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
        }
        .buttonStyle(.plain)
        .accessibilityLabel(tr("dashboard.offlineBanner", ["n": offline.count]))
    }

    private func topCpuRow(_ agent: Agent, divider: Bool) -> some View {
        let cpu = store.metricsFor(agent.id)?.cpuPercent ?? 0
        return NanoListRow(divider: divider) {
            VStack(alignment: .leading, spacing: 2) {
                NanoMono(agent.hostname, size: 14, color: t.fg, weight: .medium)
                Text("\(agent.os) · \(agent.arch)").font(.system(size: 11.5)).foregroundColor(t.fg4)
            }
        } leading: {
            Image(systemName: osIcon(agent.os)).font(.system(size: 17)).foregroundColor(t.fg3).frame(width: 20)
        } trailing: {
            HStack(spacing: 8) {
                VStack(alignment: .trailing, spacing: 3) {
                    Text(String(format: "%.0f%%", cpu)).font(.system(size: 15, weight: .semibold)).foregroundColor(t.usageColor(cpu)).monospacedDigit()
                    NanoMeter(value: cpu / 100).frame(width: 50)
                }
                Image(systemName: "chevron.right").font(.system(size: 12)).foregroundColor(t.fg4)
            }
        }
    }

    private func auditRow(_ entry: AuditEntry, divider: Bool) -> some View {
        let tone = entry.ok ? t.ok : t.crit
        let summary = entry.paramsMap.isEmpty
            ? (!entry.target.isEmpty ? entry.target : (!entry.agentHostname.isEmpty ? entry.agentHostname : entry.error))
            : entry.paramsMap.map { "\($0.key)=\($0.value)" }.sorted().joined(separator: " · ")
        return NanoListRow(divider: divider) {
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    NanoMono(entry.type.isEmpty ? "event" : entry.type, size: 13, color: t.fg, weight: .medium)
                    Text(Fmt.ago(entry.at)).font(.system(size: 11)).foregroundColor(t.fg4)
                }
                if !summary.isEmpty { NanoMono(summary, size: 11.5, color: t.fg3).lineLimit(1) }
            }
        } leading: {
            ZStack { Circle().fill(tone.opacity(0.16)); Image(systemName: auditIcon(entry.type)).font(.system(size: 14)).foregroundColor(tone) }
                .frame(width: 34, height: 34)
        } trailing: {
            if !entry.user.isEmpty { Text(entry.user).font(.system(size: 11)).foregroundColor(t.fg4) }
        }
    }

    private func emptyHint(icon: String, text: String) -> some View {
        NanoCard(padding: EdgeInsets(top: 28, leading: 12, bottom: 28, trailing: 12)) {
            VStack(spacing: 8) {
                Image(systemName: icon).font(.system(size: 26)).foregroundColor(t.fg4)
                Text(text).font(.system(size: 13)).foregroundColor(t.fg4)
            }.frame(maxWidth: .infinity)
        }
    }

    private func refresh() async {
        guard let id = store.activeServerId else { return }
        await store.fetchRecentActivity(serverId: id, limit: 50)
        pushSparks()
    }

    private func pushSparks() {
        let server = store.activeServerId
        if server != sparkServerId { sparkServerId = server; cpuSpark.removeAll(); memSpark.removeAll() }
        func push(_ buffer: inout [Double], _ value: Double) {
            if let last = buffer.last, abs(last - value) < 0.05 { return }
            buffer.append(value); if buffer.count > 20 { buffer.removeFirst() }
        }
        push(&cpuSpark, avgCpu); push(&memSpark, avgMemory)
    }
}

/// Hashable navigation wrapper while the model itself intentionally stays plain.
struct AgentRoute: Hashable {
    let agent: Agent
    static func == (lhs: AgentRoute, rhs: AgentRoute) -> Bool { lhs.agent.id == rhs.agent.id }
    func hash(into hasher: inout Hasher) { hasher.combine(agent.id) }
}

func osIcon(_ os: String) -> String {
    let s = os.lowercased()
    if s.contains("mac") || s.contains("darwin") || s.contains("ios") { return "apple.logo" }
    if s.contains("win") { return "window.ceiling" }
    return "terminal"
}

private func auditIcon(_ type: String) -> String {
    let s = type.lowercased()
    if s.contains("shell") || s.contains("exec") { return "terminal" }
    if s.contains("restart") || s.contains("reload") { return "arrow.clockwise" }
    if s.contains("reboot") || s.contains("power") || s.contains("shutdown") { return "power" }
    if s.contains("stop") || s.contains("kill") { return "stop.fill" }
    if s.contains("start") { return "play.fill" }
    return "clock.arrow.circlepath"
}
